### PR TITLE
WIP: Add download_file and URLInputFile

### PR DESCRIPTION
**Pull request by gabbhack**
# Description

- Add new download_file method to Bot
- Add new URLInputFile type

## Type of change

- [x] Documentation (typos, code examples or any documentation update)
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by gabbhack** [comment]
I have no idea how to implement Downloadable mixin correctly.

File: aiogram/api/types/mixins.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        :param timeout: Total timeout in seconds, defaults to 30
+        :type timeout: int
+        :param chunk_size: File chunks size, defaults to 64 kb
+        :type chunk_size: int
+        :param seek: Go to start of file when downloading is finished. Used only for destination with :class:`typing.BinaryIO` type, defaults to True
+        :type seek: bool
+        :param make_dirs: make dirs if not exist
+        :return: destination
+        """
+        from aiogram.api.client.bot import Bot
```
Comment (b0g3r) [inline aiogram/api/types/mixins.py]: Why local imports everywhere 🤔 

Did I miss something?

Comment (b0g3r) [reply aiogram/api/types/mixins.py]: From chat: possible circular imports

File: aiogram/api/types/mixins.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        Attention!!
+        This method has security vulnerabilities for the reason that result
+        contains bot's *access token* in open form. Use at your own risk!
+
+        :return: url
+        """
+        from aiogram.api.client.bot import Bot
+
+        file = await self.get_file()
+        return Bot.get_current(no_error=False).session.api.file_url(file.file_path)
```
Comment (b0g3r) [inline aiogram/api/types/mixins.py]: You missed the token param
https://github.com/aiogram/aiogram/blob/0bd7fc2c7e7a4e9cb9976428a6345b57ff091a4a/aiogram/api/client/telegram.py#L23

File: aiogram/api/types/mixins.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        """
+        from aiogram.api.client.bot import Bot
+
+        file = await self.get_file()
+
+        is_path = True
+        if destination is None:
+            destination = file.file_path
+        elif isinstance(destination, (str, pathlib.Path)) and os.path.isdir(destination):
+            # file from self.get_file() always have file_path
```
Comment (b0g3r) [inline aiogram/api/types/mixins.py]: Why? Docs says File object has optional file_path: https://core.telegram.org/bots/api#file

File: aiogram/api/types/mixins.py (hunk 2 above)
Comment (gabbhack) [reply aiogram/api/types/mixins.py]: Yes :(
It looks like we need to add a separate method to the Bot.

File: aiogram/api/types/mixins.py (hunk 3 above)
Comment (gabbhack) [reply aiogram/api/types/mixins.py]: This is strange. File can only be obtained from the [getFile](https://core.telegram.org/bots/api#getfile) method, the description of which says "The file can then be downloaded via the link https://api.telegram.org/file/bot<token>/<file_path>, where **<file_path> is taken from the response**".

I thought that `file_path` would be missing for a large file (more than 20MB), but in this case an error is returned "Bad Request: file is too big". So I don't know when file_path can come empty. I asked a question to the technical support of bots.

**Pull request closed without merging**